File: GaudiExamples/src/ExtendedProperties/ExtendedProperties2.cpp
/***********************************************************************************\
* (c) Copyright 1998-2019 CERN for the benefit of the LHCb and ATLAS collaborations *
*                                                                                   *
* This software is distributed under the terms of the Apache version 2 licence,     *
* copied verbatim in the file "LICENSE".                                            *
*                                                                                   *
* In applying this licence, CERN does not waive the privileges and immunities       *
* granted to it by virtue of its status as an Intergovernmental Organization        *
* or submit itself to any jurisdiction.                                             *
\***********************************************************************************/
// ============================================================================
// Include files
// ============================================================================
// STD & STL
// ============================================================================
#include <map>
#include <string>
#include <utility>
#include <vector>
// ============================================================================
// GaudiKernel
// ============================================================================
#include "GaudiKernel/GenericVectorTypes.h"
#include "GaudiKernel/Point3DTypes.h"
#include "GaudiKernel/Point4DTypes.h"
#include "GaudiKernel/Vector3DTypes.h"
#include "GaudiKernel/Vector4DTypes.h"
// ============================================================================
#include "GaudiKernel/SVectorAsProperty.h"
#include "GaudiKernel/VectorsAsProperty.h"
// ============================================================================
// GaudiAlg
// ============================================================================
#include "GaudiAlg/GaudiAlgorithm.h"
// ============================================================================

/** @file
 *  simple DEMO-file for "extended vector properties",
 *  implementation file for class Gaudi::Examples::ExtendedProperties2
 *  @author Vanya BELYAEV ibelyaev@physics.syr.edu
 *  @date  2009-09-05
 */
// ============================================================================
namespace Gaudi {
  // ==========================================================================
  namespace Examples {
    // ========================================================================
    /** @class ExtendedProperties2
     *  simple DEMO-file for "extended vector properties",
     *  @author Vanya BELYAEV ibelyaev@physics.syr.edu
     *  @date  2009-09-05
     */
    class ExtendedProperties2 : public GaudiAlgorithm {
    public:
      // ======================================================================
      StatusCode execute() override;
      // ======================================================================
    public:
      // ======================================================================
      /** standard constructor
       *  @param name algorithm instance name
       *  @param pSvc pointer to Service Locator
       */
      ExtendedProperties2( const std::string& name, ISvcLocator* pSvc ) : GaudiAlgorithm( name, pSvc ) {
        setProperty( "PropertiesPrint", true ).ignore();
      }
      // ======================================================================
    private:
      // ======================================================================
      Gaudi::Property<Gaudi::XYZPoint>               m_point3D{ this, "Point3D", { 0, 1, 2 }, "3D-point" };
      Gaudi::Property<Gaudi::XYZVector>              m_vector3D{ this, "Vector3D", { 1, 2, 3 }, "3D-vector" };
      Gaudi::Property<Gaudi::LorentzVector>          m_vector4D{ this, "Vector4D", { 1, 2, 3, 4 }, "Lorentz-vector" };
      Gaudi::Property<Gaudi::Vector5>                m_vector5{ this, "SVector5", {}, "Generic-vector" };
      Gaudi::Property<std::vector<Gaudi::XYZPoint>>  m_points{ this, "Points3D", {}, "Vector of 3D-points" };
      Gaudi::Property<std::vector<Gaudi::XYZVector>> m_vectors{ this, "Vectors3D", {}, "Vector of 3D-vectors" };
      Gaudi::Property<std::vector<Gaudi::LorentzVector>> m_lvs{ this, "Vectors4D", {}, "Vector of 4D-vectors" };
      // ======================================================================
    };
    // ========================================================================
  } // namespace Examples
  // ==========================================================================
} //                                                     end of namespace Gaudi
// ============================================================================
StatusCode Gaudi::Examples::ExtendedProperties2::execute() {
  always() << "3D-Point : " << Gaudi::Utils::toString( m_point3D.value() ) << endmsg;
  always() << "3D-Vector: " << Gaudi::Utils::toString( m_vector3D.value() ) << endmsg;
  always() << "4D-Vector: " << Gaudi::Utils::toString( m_vector4D.value() ) << endmsg;
  always() << " 5-Vector: " << Gaudi::Utils::toString( m_vector5.value() ) << endmsg;

  always() << "Vector of 3D-Points  " << Gaudi::Utils::toString( m_points.value() ) << endmsg;
  always() << "Vector of 3D-Vectors " << Gaudi::Utils::toString( m_vectors.value() ) << endmsg;
  always() << "Vector of 4D-Vectors " << Gaudi::Utils::toString( m_lvs.value() ) << endmsg;

  return StatusCode::SUCCESS;
}
// ============================================================================
// the factory
// ============================================================================
using Gaudi::Examples::ExtendedProperties2;
DECLARE_COMPONENT( ExtendedProperties2 )
// ============================================================================
// The END
// ============================================================================

File: GaudiExamples/scripts/ExtendedProperties2.py
#!/usr/bin/env python
#####################################################################################
# (c) Copyright 1998-2019 CERN for the benefit of the LHCb and ATLAS collaborations #
#                                                                                   #
# This software is distributed under the terms of the Apache version 2 licence,     #
# copied verbatim in the file "LICENSE".                                            #
#                                                                                   #
# In applying this licence, CERN does not waive the privileges and immunities       #
# granted to it by virtue of its status as an Intergovernmental Organization        #
# or submit itself to any jurisdiction.                                             #
#####################################################################################
# =============================================================================
"""
*******************************************************************************
*                                                                             *
* Simple example which illustrate the extended job-properties and their       *
*                                                                             *
*******************************************************************************
"""
from __future__ import print_function
# =============================================================================
__author__ = 'Vanya BELYAEV ibelyaev@physics.syr.edu'
# =============================================================================
# @file
#  Simple example (identical to C++ ExtendedProperties.opts) which illustrates
#  the extended job-properties and their C++/Python intercommunication
#  @author Vanya BELYAEV ibelyaev@physics.syr.edu
#  @date 2007-02-13
# =============================================================================
from Gaudi.Configuration import *

from Configurables import ApplicationMgr

from Configurables import Gaudi__Examples__ExtendedProperties2 as EP2

# Flag to check if we are running on Windows, where LorentzVectors cannot be used as properties
import sys
isWin = sys.platform.startswith("win")

ep2 = EP2("xProps2", Point3D=(10, 40, 55), Points3D=[(10, 40, 55), (1, 2, 3)])

ApplicationMgr(EvtSel="NONE", TopAlg=[ep2], EvtMax=10)

# =============================================================================
# The actual job excution
# =============================================================================
if '__main__' == __name__:

    print(__doc__, __author__)

    # make sure cling can generate all required methods in Gaudi::Property
    # Workaround for ROOT-10769
    import warnings
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        import cppyy
    for h in ("GaudiKernel/SVectorAsProperty.h",
              "GaudiKernel/VectorsAsProperty.h"):
        cppyy.gbl.gInterpreter.Declare('#include "%s"' % h)

    from GaudiPython.Bindings import AppMgr

    gaudi = AppMgr()

    gaudi.run(5)

    xp2 = gaudi.algorithm('xProps2')

    xp2.Point3D = '(-10,3, Z : 24)'
    xp2.Vector3D = [-120, -30, -40]
    if not isWin:
        xp2.Vector4D = [-100, -200, -300, 400]
        xp2.Vector4D = (-100, -200, -300, 400)
        xp2.Vector4D = [(-100, -200, -300), 400]
        xp2.Vector4D = [[-100, -200, -300], 400]
        xp2.Vector4D = ((-100, -200, -300), 400)
        xp2.Vector4D = ([-100, -200, -300], 400)

    xp2.SVector5 = (1, 2, 3, 4, 5)

    xp2.SVector5 = [1, 2, 3, 4, 5]

    try:
        xp2.SVector5 = [1, 2, 3, 4, 5, 6]
    except Exception as e:
        print(' Exception: ', e)

    try:
        xp2.Point3D = (1, 2, 3, 4)
    except Exception as e:
        print(' Exception: ', e)

    if not isWin:
        try:
            xp2.Vector4D = (1, 2, 3)
        except Exception as e:
            print(' Exception: ', e)

    xp2.Vectors3D = [(1, 2, 3), (4, 5, 6), [7, 8, 9]]
    if not isWin:
        xp2.Vectors4D = ((1, 2, 3, 4), [4, 5, 6, 7])

    xp2.PropertiesPrint = True

# =============================================================================
# The END
# =============================================================================
